datadog_checks_dev/datadog_checks/dev/config_validator/config_block.py

```diff
@@ -0,0 +1,452 @@
+# (C) Datadog, Inc. 2019
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import re
+
+from .utils import get_indent, is_at_least_indented, is_blank, is_exactly_indented
+from .validator_errors import SEVERITY_WARNING, ValidatorError
+
+# The maximum length authorized for comments. It accounts for text after '##' and does not apply to the @param
+# declaration
+MAX_COMMENT_LENGTH = 120
+
+# The list of accepted type as regex. Anything starting with list is allowed.
+ACCEPTED_VAR_TYPE_REGEX = [
+    "^boolean$",
+    "^string$",
+    "^integer$",
+    "^double$",
+    "^float$",
+    "^string$",
+    "^object$",
+    "^list.*$",
+    "^dictionary$",
+]
+
+# Regex used to parse the fields of the '## @param' declaration
+PARAM_REGEX = "^## @param ([a-zA-Z0-9_\\-]+) +- (.+) - (required|optional)( - default: .*)?$"
+
+# Regex to parse a 'key: value' item
+VAR_REGEX = "^(# |)(\\w+): ([a-zA-Z0-9_><]+)$"
+
+# Regex to identify an object (it ends with a colon). No comment allowed after the colon but the whole line
+# can be a comment.
+OBJECT_REGEX = "^(# |)([a-zA-Z0-9_\\-]+): *$"
+
+# Regex to identify a comment. Note that 'INDENT' needs to be updated with the number of spaces expected
+COMMENT_REGEX = "^ {INDENT}##(.*)$"
+
+# Regex to match any comment with no respect to indentation
+INCORRECTLY_INDENTED_COMMENT_REGEX = "^ *##(.*)$"
+
+
+class ParamProperties:
+    """Class to represent a parameter declared using the '@param' annotation
+    """
+
+    def __init__(self, var_name, type_name, required=True, default_value=None):
+        self.var_name = var_name
+        self.type_name = type_name
+        self.required = required
+        self.default_value = default_value
+
+    @staticmethod
+    def parse_from_string(idx, config_lines, indent, errors=None):
+        if errors is None:
+            errors = []
```

Comment:
That pattern is interesting. I guess you don't want to raise exceptions to accumulate errors? Did you write too much go recently? :). You'll noticed for example in this class that we only ever have one error anyway.

If you don't want to raise error, either:

- Always return 2 things, data and errors ala go. That'd be more explicit.
- Always pass `errors` (don't default to `None`). This ways there is not risk for them to be lost in the shuffle.

Reply:
Yes, the reason why I don't want to raise exceptions is that code execution does not need to be interrupted. Like we add a warning for each inline comment in the file. It's easier to append an error to an array every time we see one than to interrupt raise and catch it.

I've moved the whole code to always pass errors. I also replaced in the code any `errors += validate_something()` with `validate_something(errors)`, to make it more obvious the method does not ouput anything but updates the provided array.